File: app/src/main/java/com/daohoangson/ydls/MainActivity.java
package com.daohoangson.ydls;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;

import com.daohoangson.ydls.databinding.ActivityMainBinding;
import com.daohoangson.ydls.viewmodel.MainViewModel;
import com.google.android.gms.cast.framework.CastButtonFactory;
import com.google.android.gms.cast.framework.CastContext;
import com.google.android.gms.cast.framework.CastSession;
import com.google.android.gms.cast.framework.SessionManagerListener;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

public class MainActivity extends AppCompatActivity {

    private CastContext mCastContext;
    private SessionManagerListener<CastSession> mSessionManagerListener;

    private MainViewModel viewmodel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setupCastListener();
        mCastContext = CastContext.getSharedInstance(this);

        viewmodel = ViewModelProviders.of(this).get(MainViewModel.class);
        viewmodel.setMediaUrlFromIntent(getIntent());
        viewmodel.setCastSession(mCastContext.getSessionManager().getCurrentCastSession());

        ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setViewmodel(viewmodel);
        binding.setLifecycleOwner(this);

        viewmodel.getActionStartActivity().observe(this, new Observer<Class>() {
            @Override
            public void onChanged(Class aClass) {
                Intent intent = new Intent(MainActivity.this, aClass);
                startActivity(intent);
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);

        getMenuInflater().inflate(R.menu.cast, menu);
        CastButtonFactory.setUpMediaRouteButton(getApplicationContext(), menu, R.id.media_route_menu_item);

        return true;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        viewmodel.setMediaUrlFromIntent(intent);
    }

    @Override
    protected void onPause() {
        super.onPause();

        mCastContext.getSessionManager().removeSessionManagerListener(mSessionManagerListener, CastSession.class);
    }

    @Override
    protected void onResume() {
        mCastContext.getSessionManager().addSessionManagerListener(mSessionManagerListener, CastSession.class);

        super.onResume();
    }

    private void setupCastListener() {
        mSessionManagerListener = new SessionManagerListener<CastSession>() {

            @Override
            public void onSessionEnded(CastSession session, int error) {
                onApplicationDisconnected();
            }

            @Override
            public void onSessionResumed(CastSession session, boolean wasSuspended) {
                onApplicationConnected(session);
            }

            @Override
            public void onSessionResumeFailed(CastSession session, int error) {
                onApplicationDisconnected();
            }

            @Override
            public void onSessionStarted(CastSession session, String sessionId) {
                onApplicationConnected(session);
            }

            @Override
            public void onSessionStartFailed(CastSession session, int error) {
                onApplicationDisconnected();
            }

            @Override
            public void onSessionStarting(CastSession session) {
            }

            @Override
            public void onSessionEnding(CastSession session) {
            }

            @Override
            public void onSessionResuming(CastSession session, String sessionId) {
            }

            @Override
            public void onSessionSuspended(CastSession session, int reason) {
            }

            private void onApplicationConnected(CastSession session) {
                viewmodel.setCastSession(session);
            }

            private void onApplicationDisconnected() {
                viewmodel.setCastSession(null);
            }
        };
    }
}

File: app/src/main/java/com/daohoangson/ydls/MainActivity.kt
package com.daohoangson.ydls

import android.content.Intent
import android.os.Bundle
import android.view.Menu

import com.daohoangson.ydls.databinding.ActivityMainBinding
import com.daohoangson.ydls.viewmodel.MainViewModel
import com.google.android.gms.cast.framework.CastButtonFactory
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.CastSession
import com.google.android.gms.cast.framework.SessionManagerListener

import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders

class MainActivity : AppCompatActivity() {

    private var mCastContext: CastContext? = null
    private var mSessionManagerListener: SessionManagerListener<CastSession>? = null

    private var viewmodel: MainViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupCastListener()
        mCastContext = CastContext.getSharedInstance(this)

        viewmodel = ViewModelProviders.of(this).get(MainViewModel::class.java)
        viewmodel!!.setMediaUrlFromIntent(intent)
        viewmodel!!.setCastSession(mCastContext!!.sessionManager.currentCastSession)

        val binding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        binding.viewmodel = viewmodel
        binding.setLifecycleOwner(this)

        viewmodel!!.actionStartActivity.observe(this, Observer { aClass ->
            val intent = Intent(this@MainActivity, aClass)
            startActivity(intent)
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)

        menuInflater.inflate(R.menu.cast, menu)
        CastButtonFactory.setUpMediaRouteButton(applicationContext, menu, R.id.media_route_menu_item)

        return true
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        viewmodel!!.setMediaUrlFromIntent(intent)
    }

    override fun onPause() {
        super.onPause()

        mCastContext!!.sessionManager.removeSessionManagerListener(mSessionManagerListener, CastSession::class.java)
    }

    override fun onResume() {
        mCastContext!!.sessionManager.addSessionManagerListener(mSessionManagerListener!!, CastSession::class.java)

        super.onResume()
    }

    private fun setupCastListener() {
        mSessionManagerListener = object : SessionManagerListener<CastSession> {

            override fun onSessionEnded(session: CastSession, error: Int) {
                onApplicationDisconnected()
            }

            override fun onSessionResumed(session: CastSession, wasSuspended: Boolean) {
                onApplicationConnected(session)
            }

            override fun onSessionResumeFailed(session: CastSession, error: Int) {
                onApplicationDisconnected()
            }

            override fun onSessionStarted(session: CastSession, sessionId: String) {
                onApplicationConnected(session)
            }

            override fun onSessionStartFailed(session: CastSession, error: Int) {
                onApplicationDisconnected()
            }

            override fun onSessionStarting(session: CastSession) {}

            override fun onSessionEnding(session: CastSession) {}

            override fun onSessionResuming(session: CastSession, sessionId: String) {}

            override fun onSessionSuspended(session: CastSession, reason: Int) {}

            private fun onApplicationConnected(session: CastSession) {
                viewmodel!!.setCastSession(session)
            }

            private fun onApplicationDisconnected() {
                viewmodel!!.setCastSession(null)
            }
        }
    }
}
